src/linux/mod.rs

```diff
@@ -238,23 +247,33 @@ impl<'a> SharedLibraryTrait for SharedLibrary<'a> {
                     continue;
                 }
 
-                let mut offset = phdr.p_offset as usize;
-                let end = offset + phdr.p_filesz as usize;
+                let mut note_header_addr = self.program_header_vm_addr(phdr) as usize;
+                let end_of_note = (note_header_addr as usize) + (phdr.p_memsz as usize);
 
-                while offset < end {
-                    // we always use an nhdr32 here as 64bit notes have not
+                while note_header_addr < end_of_note {
+                    // Each entry in a `PT_NOTE` segment begins with a
+                    // fixed-size header `Nhdr`.
+                    //
+                    // We always use an Nhdr32 here as 64-bit notes have not
                     // been observed in practice.
-                    let nhdr = &*((self.addr as usize + offset) as *const Nhdr32);
-                    offset += mem::size_of_val(nhdr);
-                    offset += nhdr.n_namesz as usize;
-                    align(alignment, &mut offset);
-                    let value =
-                        slice::from_raw_parts(self.addr.add(offset), nhdr.n_descsz as usize);
-                    offset += nhdr.n_descsz as usize;
-                    align(alignment, &mut offset);
+                    let nhdr = (note_header_addr as *const Nhdr32).as_ref().unwrap();
```

Comment:
This doesn't check that there are still enough bytes left for the header (and similar for name/desc below). Not sure how pedantic we need to be here, but this is the kind of thing that would be easier to check if we used a slice.